### PR TITLE
Starting to tidy up mathics.core.pattern

### PR DESCRIPTION
**Pull request by mmatera**
In line with the last changes proposed by @rocky, I was doing a pass over mathics.core.pattern and fixing some issues reported by the linter.

### CODE REVIEW

File: mathics/core/pattern.py [hunk 1]
```diff
@@ -1,6 +1,11 @@
 # cython: language_level=3
```
Comment (rocky) [inline mathics/core/pattern.py:1]: BTW - I think we should remove cython. At this point it is probably not speeding things up and getting in the way.

File: mathics/core/pattern.py [hunk 2]
```diff
@@ -906,9 +822,197 @@ def get_match_candidates_count(
 
         count = 0
         for element in elements:
-            if self.does_match(element, evaluation, vars):
+            if self.does_match(element, evaluation, vars_dict):
                 count += 1
         return count
 
     def sort(self):
+        """Sot the elements according to their sort key"""
```
Comment (rocky) [inline mathics/core/pattern.py:830]: Sot the -> Sort

File: mathics/core/pattern.py (hunk 1 above)
Comment (mmatera) [reply mathics/core/pattern.py:1]: For this, Cython was useful in detecting some errors when I renamed some arguments. Maybe we can remove this later.

**Comment by rocky** [comment]
As I look over I see lots of opportunity for adding type annotations. But one step at a time. See comments and merge when you are satisfied. 

Also consider making a pass over https://github.com/Mathics3/mathics-development-guide/blob/master/docs/extending/code-overview/pattern-matching.rst (but note that this section is going to be moved up a level out of "Extending" and its own section.

File: mathics/core/pattern.py (hunk 1 above)
Comment (rocky) [reply mathics/core/pattern.py:1]: Ok - leave in then.

**Comment by mmatera** [comment]
> As I look over I see lots of opportunity for adding type annotations. But one step at a time. See comments and merge when you are satisfied.
> 

Fist, I need to be sure about what are the parameters that the functions and methods are receiving. Also, one of the complaints of the linter is that many of these functions have too many parameters. Probably a first step would be to collect all these parameters into a dictionary or a tuple, and then establish the typing.


> Also consider making a pass over https://github.com/Mathics3/mathics-development-guide/blob/master/docs/extending/code-overview/pattern-matching.rst (but note that this section is going to be moved up a level out of "Extending" and its own section.


OK

**Comment by rocky** [comment]
Concerning the https://github.com/Mathics3/mathics-core/actions/runs/10905508459/job/30264517603?pr=1086 failure, I suspect that various linters detect this too. In other

**Comment by rocky** [comment]
@mmatera When you think things are stable, let me know. I'd like to checkout this branch and look it over to see if this helps me understand how patterns work better. Thanks.

**Comment by mmatera** [comment]
> @mmatera When you think things are stable, let me know. I'd like to checkout this branch and look it over to see if this helps me understand how patterns work better. Thanks.

I am nearly to finish: I am trying to determine the right annotations using the errors reported by Cython. I'll let you when I finish

**Comment by mmatera** [comment]
@rocky, I think this is ready now. In a next PR I will go over `mathics.builtin.patterns.py` doing the same.

**Comment by rocky** [comment]
> @rocky, I think this is ready now. In a next PR I will go over `mathics.builtin.patterns.py` doing the same.
 
Thanks - give me another day or so to try this out.

**Comment by rocky** [comment]
LGTM  - let's merge and iterate. The developer docs and debugger will now need to track these changes.

**Pull request merged**